buddy_alloc.h

```diff
@@ -840,9 +932,9 @@ void *buddy_walk(struct buddy *buddy,
             }
 
             // Current node is allocated, process
-            size_t pos_size = effective_memory_size >> (state.current_pos.depth - 1u);
-            unsigned char *addr = address_for_position(buddy, state.current_pos);
-            if (((size_t)(addr - main + pos_size)) > buddy->memory_size) {
+            pos_size = effective_memory_size >> (state.current_pos.depth - 1u);
+            addr = address_for_position(buddy, state.current_pos);
+            if (((size_t)((uintptr_t)addr - (uintptr_t)main + pos_size)) > buddy->memory_size) {
```

Comment:
addr - main is ptrdiff_t type, when this adds with pos_size which is size_t (unsigned). Somehow convert the pointer to uintptr_t silent the warning. Is there any guarantee that addr-main+pos is positive? Otherwise, the result convert to size_t would have a underflow. Also, looking at this code, is there any guarantee that current_pos.depth >= 1?

Reply:
See `test_buddy_tree_depth` in tests.c about the depth of the pos structure. For any valid position the depth is always a positive, non-zero number.

Reply:
The addressing here is correct as well - addr will always be the same as main or higher and pos_size is the positive length of the allocated slot.